Problems writing a Injection for a refined type

Hi all,
I'm trying to write an `Injection` that converts back and forth between a [refined type](https://github.com/fthomas/refined) and a primitive type:
https://gist.github.com/nightscape/adba260f7f7a3943417858004cd70d43
However, I'm running into an exception which seems to happen before the `Injection` is even applied:
```
Caused by: java.lang.ClassCastException: java.lang.Integer cannot be cast to eu.timepit.refined.api.Refined
        at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
        at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:287)
        at org.apache.spark.sql.SparkSession$$anonfun$2.apply(SparkSession.scala:464)
        at org.apache.spark.sql.SparkSession$$anonfun$2.apply(SparkSession.scala:464)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
        at scala.collection.immutable.List.map(List.scala:296)
        at org.apache.spark.sql.SparkSession.createDataset(SparkSession.scala:464)
        at org.apache.spark.sql.SQLContext.createDataset(SQLContext.scala:377)
        at frameless.TypedDataset$.create(TypedDataset.scala:637)
```

The problem looks somewhat similar to https://github.com/typelevel/frameless/issues/102 but there's no `Vector`s involved.
Do you have any idea what could be going wrong here?

Thanks a lot!

--- a/build.sbt
+++ b/build.sbt
@@ -7,6 +7,7 @@ val scalatestplus = "3.1.0.0-RC2"
 val shapeless = "2.3.7"
 val scalacheck = "1.15.4"
 val irrecVersion = "0.4.0"
+val refinedVersion = "0.9.27"
 
 val Scala212 = "2.12.15"
 
@@ -20,21 +21,21 @@ ThisBuild / scalaVersion := (ThisBuild / crossScalaVersions).value.last
 ThisBuild / mimaFailOnNoPrevious := false
 
 lazy val root = Project("frameless", file("." + "frameless")).in(file("."))
-  .aggregate(core, cats, dataset, ml, docs)
-  .settings(framelessSettings: _*)
-  .settings(noPublishSettings: _*)
-  .settings(mimaPreviousArtifacts := Set())
+  .aggregate(core, cats, dataset, refined, ml, docs)
+  .settings(framelessSettings)
+  .settings(noPublishSettings)
+  .settings(mimaPreviousArtifacts := Set.empty)
 
 lazy val core = project
   .settings(name := "frameless-core")
-  .settings(framelessSettings: _*)
-  .settings(publishSettings: _*)
+  .settings(framelessSettings)
+  .settings(publishSettings)
 
 
 lazy val cats = project
   .settings(name := "frameless-cats")
-  .settings(framelessSettings: _*)
-  .settings(publishSettings: _*)
+  .settings(framelessSettings)
+  .settings(publishSettings)
   .settings(
     addCompilerPlugin("org.typelevel" % "kind-projector" % "0.13.2" cross CrossVersion.full),
     scalacOptions += "-Ypartial-unification"
@@ -57,7 +58,7 @@ lazy val dataset = project
     libraryDependencies ++= Seq(
       "org.apache.spark" %% "spark-core"      % sparkVersion % Provided,
       "org.apache.spark" %% "spark-sql"       % sparkVersion % Provided,
-      "net.ceedubs"      %% "irrec-regex-gen" % irrecVersion % Test
+      "net.ceedubs"      %% "irrec-regex-gen" % irrecVersion % Test,
     ),
     mimaBinaryIssueFilters ++= {
       import com.typesafe.tools.mima.core._
@@ -79,6 +80,21 @@ lazy val dataset = project
     }
   ))
   .dependsOn(core % "test->test;compile->compile")
+
+lazy val refined = project
+  .settings(name := "frameless-refined")
+  .settings(framelessSettings)
+  .settings(framelessTypedDatasetREPL)
+  .settings(publishSettings)
+  .settings(Seq(
+    mimaPreviousArtifacts := Set.empty,
+    libraryDependencies ++= Seq(
+      "org.apache.spark" %% "spark-core"      % sparkVersion % Provided,
+      "org.apache.spark" %% "spark-sql"       % sparkVersion % Provided,
+      "eu.timepit"       %% "refined"         % refinedVersion
+    )
+  ))
+  .dependsOn(dataset % "test->test;compile->compile")
 
 lazy val ml = project
   .settings(name := "frameless-ml")

--- a/dataset/src/main/scala/frameless/RecordEncoder.scala
+++ b/dataset/src/main/scala/frameless/RecordEncoder.scala
@@ -1,6 +1,7 @@
 package frameless
 
 import org.apache.spark.sql.FramelessInternals
+
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.objects.{
   Invoke, NewInstance, UnwrapOption, WrapOption
@@ -22,6 +23,9 @@ case class RecordEncoderField(
 
 trait RecordEncoderFields[T <: HList] extends Serializable {
   def value: List[RecordEncoderField]
+
+  override def toString: String =
+    s"""RecordEncoderFields${value.mkString("[", ", ", "]")}"""
 }
 
 object RecordEncoderFields {
@@ -164,11 +168,13 @@ class RecordEncoder[F, G <: HList, H <: HList]
 
     def fromCatalyst(path: Expression): Expression = {
       val exprs = fields.value.value.map { field =>
-        field.encoder.fromCatalyst( GetStructField(path, field.ordinal, Some(field.name)) )
+        field.encoder.fromCatalyst(
+          GetStructField(path, field.ordinal, Some(field.name)))
       }
 
       val newArgs = newInstanceExprs.value.from(exprs)
-      val newExpr = NewInstance(classTag.runtimeClass, newArgs, jvmRepr, propagateNull = true)
+      val newExpr = NewInstance(
+        classTag.runtimeClass, newArgs, jvmRepr, propagateNull = true)
 
       val nullExpr = Literal.create(null, jvmRepr)
 

--- a/dataset/src/main/scala/frameless/TypedEncoder.scala
+++ b/dataset/src/main/scala/frameless/TypedEncoder.scala
@@ -53,6 +53,8 @@ object TypedEncoder {
 
     def fromCatalyst(path: Expression): Expression =
       Invoke(path, "toString", jvmRepr)
+
+    override val toString = "stringEncoder"
   }
 
   implicit val booleanEncoder: TypedEncoder[Boolean] =
@@ -74,6 +76,8 @@ object TypedEncoder {
 
     def toCatalyst(path: Expression): Expression = path
     def fromCatalyst(path: Expression): Expression = path
+
+    override def toString = "intEncoder"
   }
 
   implicit val longEncoder: TypedEncoder[Long] = new TypedEncoder[Long] {

--- a/refined/src/main/scala/frameless/refined/RefinedFieldEncoders.scala
+++ b/refined/src/main/scala/frameless/refined/RefinedFieldEncoders.scala
@@ -1,0 +1,79 @@
+package frameless.refined
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.objects.{
+  Invoke, NewInstance, UnwrapOption, WrapOption
+}
+import org.apache.spark.sql.types._
+
+import eu.timepit.refined.api.RefType
+
+import frameless.{ TypedEncoder, RecordFieldEncoder }
+
+private[refined] trait RefinedFieldEncoders {
+  /**
+   * @tparam T the refined type (e.g. `String`)
+   */
+  implicit def optionRefined[F[_, _], T, R](
+    implicit
+      i0: RefType[F],
+      i1: TypedEncoder[T],
+      i2: ClassTag[F[T, R]],
+  ): RecordFieldEncoder[Option[F[T, R]]] =
+    RecordFieldEncoder[Option[F[T, R]]](new TypedEncoder[Option[F[T, R]]] {
+      def nullable = true
+
+      // `Refined` is a Value class: https://github.com/fthomas/refined/blob/master/modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/api/Refined.scala#L8
+      def jvmRepr = ObjectType(classOf[Option[F[T, R]]])
+
+      def catalystRepr: DataType = i1.catalystRepr
+
+      val innerJvmRepr = ObjectType(i2.runtimeClass)
+
+      def fromCatalyst(path: Expression): Expression = {
+        val javaValue = i1.fromCatalyst(path)
+        val value = NewInstance(i2.runtimeClass, Seq(javaValue), innerJvmRepr)
+
+        WrapOption(value, innerJvmRepr)
+      }
+
+      @inline def toCatalyst(path: Expression): Expression = {
+        val value = UnwrapOption(innerJvmRepr, path)
+
+        val javaValue = Invoke(value, "value", i1.jvmRepr, Nil)
+
+        i1.toCatalyst(javaValue)
+      }
+
+      override def toString = s"optionRefined[${i2.runtimeClass.getName}]"
+    })
+
+  /**
+   * @tparam T the refined type (e.g. `String`)
+   */
+  implicit def refined[F[_, _], T, R](
+    implicit
+      i0: RefType[F],
+      i1: TypedEncoder[T],
+      i2: ClassTag[F[T, R]],
+  ): RecordFieldEncoder[F[T, R]] =
+    RecordFieldEncoder[F[T, R]](new TypedEncoder[F[T, R]] {
+      def nullable = i1.nullable
+
+      // `Refined` is a Value class: https://github.com/fthomas/refined/blob/master/modules/core/shared/src/main/scala-3.0-/eu/timepit/refined/api/Refined.scala#L8
+      def jvmRepr = i1.jvmRepr
+
+      def catalystRepr: DataType = i1.catalystRepr
+
+      def fromCatalyst(path: Expression): Expression =
+        i1.fromCatalyst(path)
+
+      @inline def toCatalyst(path: Expression): Expression =
+        i1.toCatalyst(path)
+
+      override def toString = s"refined[${i2.runtimeClass.getName}]"
+    })
+}
+

--- a/refined/src/main/scala/frameless/refined/package.scala
+++ b/refined/src/main/scala/frameless/refined/package.scala
@@ -1,0 +1,33 @@
+package frameless
+
+import scala.reflect.ClassTag
+
+import eu.timepit.refined.api.{ RefType, Validate }
+
+package object refined extends RefinedFieldEncoders {
+  implicit def refinedInjection[F[_, _], T, R](
+    implicit
+      refType: RefType[F],
+      validate: Validate[T, R]
+    ): Injection[F[T, R], T] = Injection(
+    refType.unwrap,
+    { value =>
+      refType.refine[R](value) match {
+        case Left(errMsg) =>
+          throw new IllegalArgumentException(
+            s"Value $value does not satisfy refinement predicate: $errMsg")
+
+        case Right(res) => res
+      }
+    })
+
+  implicit def refinedEncoder[F[_, _], T, R](
+    implicit
+      i0: RefType[F],
+      i1: Validate[T, R],
+      i2: TypedEncoder[T],
+      i3: ClassTag[F[T, R]]
+    ): TypedEncoder[F[T, R]] = TypedEncoder.usingInjection(
+    i3, refinedInjection, i2)
+}
+

--- a/refined/src/test/scala/frameless/RefinedFieldEncoderTests.scala
+++ b/refined/src/test/scala/frameless/RefinedFieldEncoderTests.scala
@@ -1,0 +1,120 @@
+package frameless
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.{
+  IntegerType, ObjectType, StringType, StructField, StructType
+}
+
+import org.scalatest.matchers.should.Matchers
+
+class RefinedFieldEncoderTests extends TypedDatasetSuite with Matchers {
+  test("Encode a bare refined type") {
+    import eu.timepit.refined.auto._
+    import eu.timepit.refined.types.string.NonEmptyString
+
+    val encoder: TypedEncoder[NonEmptyString] = {
+      import frameless.refined.refinedEncoder
+      TypedEncoder[NonEmptyString]
+    }
+
+    val ss = session
+    import ss.implicits._
+
+    encoder.catalystRepr shouldBe StringType
+
+    val nes: NonEmptyString = "Non Empty String"
+
+    val unsafeDs = TypedDataset.createUnsafe(sc.parallelize(Seq(nes.value)).toDF())(encoder)
+
+    val expected = Seq(nes)
+
+    unsafeDs.collect.run() shouldBe expected
+  }
+
+  test("Encode case class with a refined field") {
+    import RefinedTypesTests._
+
+    // Check jvmRepr
+    import org.apache.spark.sql.types.ObjectType
+
+    encoderA.jvmRepr shouldBe ObjectType(classOf[A])
+
+    // Check catalystRepr
+    val expectedAStructType = StructType(Seq(
+      StructField("a", IntegerType, false),
+      StructField("s", StringType, false)))
+
+    encoderA.catalystRepr shouldBe expectedAStructType
+
+    // Check unsafe
+    val unsafeDs: TypedDataset[A] = {
+      val rdd = sc.parallelize(Seq(Row(as.a, as.s.toString)))
+      val df = session.createDataFrame(rdd, expectedAStructType)
+
+      TypedDataset.createUnsafe(df)(encoderA)
+    }
+
+    val expected = Seq(as)
+
+    unsafeDs.collect.run() shouldBe expected
+
+    // Check safe
+    val safeDs = TypedDataset.create(expected)
+
+    safeDs.collect.run() shouldBe expected
+  }
+
+  test("Encode case class with a refined optional field") {
+    import RefinedTypesTests._
+
+    // Check jvmRepr
+    encoderB.jvmRepr shouldBe ObjectType(classOf[B])
+
+    // Check catalystRepr
+    val expectedBStructType = StructType(Seq(
+      StructField("a", IntegerType, false),
+      StructField("s", StringType, true)))
+
+    encoderB.catalystRepr shouldBe expectedBStructType
+
+    // Check unsafe
+    val unsafeDs: TypedDataset[B] = {
+      val rdd = sc.parallelize(Seq(
+        Row(bs.a, bs.s.mkString),
+        Row(2, null.asInstanceOf[String]),
+      ))
+
+      val df = session.createDataFrame(rdd, expectedBStructType)
+
+      TypedDataset.createUnsafe(df)(encoderB)
+    }
+
+    val expected = Seq(bs, B(2, None))
+
+    unsafeDs.collect.run() shouldBe expected
+
+    // Check safe
+    val safeDs = TypedDataset.create(expected)
+
+    safeDs.collect.run() shouldBe expected
+  }
+}
+
+object RefinedTypesTests {
+  import eu.timepit.refined.auto._
+  import eu.timepit.refined.types.string.NonEmptyString
+
+  case class A(a: Int, s: NonEmptyString)
+  case class B(a: Int, s: Option[NonEmptyString])
+
+  val nes: NonEmptyString = "Non Empty String"
+
+  val as = A(-42, nes)
+  val bs = B(-42, Option(nes))
+
+  import frameless.refined._ // implicit instances for refined
+
+  implicit val encoderA: TypedEncoder[A] = TypedEncoder.usingDerivation
+
+  implicit val encoderB: TypedEncoder[B] = TypedEncoder.usingDerivation
+}

